HBase connection issue Gen2 Datalake

Currently, when connecting to HBase the following error is observed:

```
Caused by: java.lang.ExceptionInInitializerError
        at org.apache.hadoop.hbase.ClusterId.parseFrom(ClusterId.java:64)
        at org.apache.hadoop.hbase.zookeeper.ZKClusterId.readClusterIdZNode(ZKClusterId.java:75)
        at org.apache.hadoop.hbase.client.ZooKeeperRegistry.getClusterId(ZooKeeperRegistry.java:105)
        at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.retrieveClusterId(ConnectionManager.java:889)
        at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.<init>(ConnectionManager.java:645)
        ... 69 more
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem not found
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2228)
        at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2780)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2793)
        at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:99)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2829)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2811)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:390)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:179)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:374)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)
        at org.apache.hadoop.hbase.util.DynamicClassLoader.initTempDir(DynamicClassLoader.java:118)
        at org.apache.hadoop.hbase.util.DynamicClassLoader.<init>(DynamicClassLoader.java:98)
        at org.apache.hadoop.hbase.protobuf.ProtobufUtil.<clinit>(ProtobufUtil.java:250)
        ... 74 more
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem not found
        at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2134)
        at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2226)
        ... 86 more
```

On the second attempt to connect to Hbase a similar error occurs but listing a missing `protobuf` class

By updating the `pom` to reference a version of the hadoop jar which we know to contain the `azureBlobFilesystem` class ([done here](https://github.com/opencb/opencga/pull/1070/commits/d39ec730520f6ba9bf153e078221fe34a6bd2ac6)] this error changed to be the following:

```
WARN  Configuration:2731 - hbase-site.xml:an attempt to override final parameter: dfs.support.append;  Ignoring.
2019-01-24 12:49:44 [http-nio-8080-exec-6] ERROR OpenCGAWSServer:447 - Catch error: Problems opening connection to DB
java.lang.IllegalStateException: Problems opening connection to DB
	at org.opencb.opencga.storage.hadoop.utils.HBaseManager.getConnection(HBaseManager.java:131)
	at org.opencb.opencga.storage.hadoop.utils.HBaseManager.<init>(HBaseManager.java:79)
	at org.opencb.opencga.storage.hadoop.variant.metadata.AbstractHBaseDBAdaptor.<init>(AbstractHBaseDBAdaptor.java:48)
	at org.opencb.opencga.storage.hadoop.variant.metadata.HBaseProjectMetadataDBAdaptor.<init>(HBaseProjectMetadataDBAdaptor.java:46)
	at org.opencb.opencga.storage.hadoop.variant.metadata.HBaseVariantStorageMetadataDBAdaptorFactory.buildProjectMetadataDBAdaptor(HBaseVariantStorageMetadataDBAdaptorFactory.java:38)
	at org.opencb.opencga.storage.hadoop.variant.metadata.HBaseVariantStorageMetadataDBAdaptorFactory.buildProjectMetadataDBAdaptor(HBaseVariantStorageMetadataDBAdaptorFactory.java:13)
	at org.opencb.opencga.storage.core.metadata.StudyConfigurationManager.<init>(StudyConfigurationManager.java:76)
	at org.opencb.opencga.storage.hadoop.variant.HadoopVariantStorageEngine.getStudyConfigurationManager(HadoopVariantStorageEngine.java:923)
	at org.opencb.opencga.storage.core.manager.variant.VariantStorageManager.secure(VariantStorageManager.java:512)
	at org.opencb.opencga.storage.core.manager.variant.VariantStorageManager.get(VariantStorageManager.java:352)
	at org.opencb.opencga.server.rest.analysis.VariantAnalysisWSService.getVariants(VariantAnalysisWSService.java:350)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.opencb.opencga.server.CORSFilter.doFilter(CORSFilter.java:46)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:493)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:650)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:800)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: java.lang.reflect.InvocationTargetException
	at org.apache.hadoop.hbase.client.ConnectionFactory.createConnection(ConnectionFactory.java:240)
	at org.apache.hadoop.hbase.client.ConnectionFactory.createConnection(ConnectionFactory.java:218)
	at org.apache.hadoop.hbase.client.ConnectionFactory.createConnection(ConnectionFactory.java:119)
	at org.opencb.opencga.storage.hadoop.utils.HBaseManager.getConnection(HBaseManager.java:125)
	... 61 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.hadoop.hbase.client.ConnectionFactory.createConnection(ConnectionFactory.java:238)
	... 64 more
Caused by: java.lang.NoClassDefFoundError: org/apache/hadoop/fs/StreamCapabilities
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at org.apache.catalina.loader.WebappClassLoaderBase.findClassInternal(WebappClassLoaderBase.java:2356)
	at org.apache.catalina.loader.WebappClassLoaderBase.findClass(WebappClassLoaderBase.java:830)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1297)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1156)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.initialize(AzureBlobFileSystem.java:103)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2795)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:99)
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2829)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2811)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:390)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:179)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:374)
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)
	at org.apache.hadoop.hbase.util.DynamicClassLoader.initTempDir(DynamicClassLoader.java:118)
	at org.apache.hadoop.hbase.util.DynamicClassLoader.<init>(DynamicClassLoader.java:98)
	at org.apache.hadoop.hbase.protobuf.ProtobufUtil.<clinit>(ProtobufUtil.java:250)
	at org.apache.hadoop.hbase.ClusterId.parseFrom(ClusterId.java:64)
	at org.apache.hadoop.hbase.zookeeper.ZKClusterId.readClusterIdZNode(ZKClusterId.java:75)
	at org.apache.hadoop.hbase.client.ZooKeeperRegistry.getClusterId(ZooKeeperRegistry.java:105)
	at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.retrieveClusterId(ConnectionManager.java:889)
	at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.<init>(ConnectionManager.java:645)
	... 69 more
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.fs.StreamCapabilities
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1328)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1156)
	... 93 more
```

My current thinking is that the version used for all of the hadoop dependencies needs be updated. 

Currently the way we're identifying which version of the hortonworks jars contain the class we need is with the following:

```
~/source/hacks/opencga-azure/arm [lg/azure-fitness L|✚ 1] 
17:24 $ jar -tf hadoop-azure-3.1.1.3.0.2.0-50.jar | grep AzureBlob
org/apache/hadoop/fs/azurebfs/contracts/exceptions/AzureBlobFileSystemException.class
org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem$1.class
org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem$2.class
org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem$FileSystemOperation.class
org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.class
org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore$VersionedFileStatus.class
org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.class
org/apache/hadoop/fs/azurebfs/SecureAzureBlobFileSystem.class
```

So I have a new error now. Having read about how `tomcat` finds stuff I realized that `classpath` won't have the effect needed [blog post here](https://www.mulesoft.com/tcat/tomcat-classpath)

Given that I then edited `catalina.properties`: ` /opt/tomcat/conf/catalina.properties` and added 

```
common.loader="${catalina.base}/lib","${catalina.base}/lib/*.jar","${catalina.home}/lib","${catalina.home}/lib/*.jar","/opt/opencga/conf/hadoop/*.jar"
```

The new error I get is the following, this suggests that I've successfully forced loading of the azure jars from the cluster which are copied into `/opt/opencga/conf/hadoop` but there is a version mismatch somewhere

```
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.apache.hadoop.hbase.protobuf.ProtobufUtil
	at org.apache.hadoop.hbase.ClusterId.parseFrom(ClusterId.java:64)
	at org.apache.hadoop.hbase.zookeeper.ZKClusterId.readClusterIdZNode(ZKClusterId.java:75)
	at org.apache.hadoop.hbase.client.ZooKeeperRegistry.getClusterId(ZooKeeperRegistry.java:105)
	at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.retrieveClusterId(ConnectionManager.java:889)
	at org.apache.hadoop.hbase.client.ConnectionManager$HConnectionImplementation.<init>(ConnectionManager.java:645)
	... 69 more
```

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -16,7 +16,7 @@ pipeline {
                 timeout(time: 30, unit: 'MINUTES')
             }
             steps {
-                sh 'mvn clean install -DskipTests -Dstorage-hadoop -Popencga-storage-hadoop-deps -Phdp-2.6.0 -DOPENCGA.STORAGE.DEFAULT_ENGINE=hadoop -Dopencga.war.name=opencga -Dcheckstyle.skip'
+                sh 'mvn clean install -DskipTests -Dstorage-mongodb -Dstorage-hadoop -Popencga-storage-hadoop-deps -Phdp-2.6.5 -DOPENCGA.STORAGE.DEFAULT_ENGINE=hadoop -Dopencga.war.name=opencga -Dcheckstyle.skip'
             }
         }
 

--- a/opencga-app/app/scripts/docker/build-all.sh
+++ b/opencga-app/app/scripts/docker/build-all.sh
@@ -42,6 +42,7 @@
 # - DOCKER_BUILD_ARGS='' : (optional) Additional build arguments to pass to the docker build command
 # - PUBLISH=''           : (optional) Set to 'true' to publish the docker images to a container registry
 # - TAG=''               : (optional) Set to override the default Git commit SHA docker image tag
+# - HADOOP_FLAVOR='hdp-2.6.5' : (optional) Compile against an specific Hadoop flavour and version
 #
 # When you have set your desired variables, you can simply run the Makefile with `make`.
 
@@ -108,13 +109,14 @@ ENVFILE="${envfile}" make --no-print-directory -f "${dockerDir}/Makefile.docker"
 
 # Build OpenCGA
 if [ ! -d "${buildPath}" ]; then
+    HADOOP_FLAVOR=${HADOOP_FLAVOR:-hdp-2.6.5}
     echo "> No existing OpenCGA build, building from source..."
     docker run -it --rm \
     -v "$PWD":/src \
     -v "$HOME/.m2":/root/.m2 \
     -w /src maven:3.6-jdk-8 \
     mvn -T 1C install \
-    -DskipTests -Dstorage-hadoop -Popencga-storage-hadoop-deps -Phdp-2.6.0 -DOPENCGA.STORAGE.DEFAULT_ENGINE=hadoop -Dopencga.war.name=opencga
+    -DskipTests -Dstorage-hadoop -Popencga-storage-hadoop-deps -P"${HADOOP_FLAVOR}" -DOPENCGA.STORAGE.DEFAULT_ENGINE=hadoop -Dopencga.war.name=opencga
     echo "> Finished OpenCGA build"
 else
     echo "> Using existing OpenCGA build from $PWD/build"

--- a/opencga-storage/opencga-storage-hadoop/pom.xml
+++ b/opencga-storage/opencga-storage-hadoop/pom.xml
@@ -237,7 +237,7 @@
             </activation>
             <properties>
                 <opencga-storage-hadoop-deps.classifier>hdp-2.6.5</opencga-storage-hadoop-deps.classifier>
-                <hdp.dependencies.version>2.6.5.90-1</hdp.dependencies.version>
+                <hdp.dependencies.version>2.6.5.3007-2</hdp.dependencies.version>
 
                 <hadoop.version>2.7.3.${hdp.dependencies.version}</hadoop.version>
                 <hbase.version>1.1.2.${hdp.dependencies.version}</hbase.version>

